### PR TITLE
Add assigned build ID to Visibility

### PR DESCRIPTION
**Pull request by ShahabT**
## What changed?

Adding the assigned build ID to BuildIds SA as `"assigned:<bld>"` in addition to existing `"versioned:<bld>"`.

## Why?

Reachability needs to query based on the currently  assigned build ID.

## How did you test it?

Added SA validation to all functional tests.

## Potential risks

None.

## Documentation

None.

## Is hotfix candidate?

No

### CODE REVIEW

File: tests/versioning.go [hunk 1]
```diff
@@ -3736,15 +3741,24 @@ func (s *VersioningIntegSuite) validateWorkflowBuildId(
...
 	expectingStamp bool,
 	expectedInheritedBuildId string,
+	extraSearchAttrBuildIds []string,
 ) {
 	dw, err := s.sdkClient.DescribeWorkflowExecution(ctx, wfId, runId)
 	s.NoError(err)
+	var searchAttr []string
+	err = converter.GetDefaultDataConverter().FromPayload(dw.GetWorkflowExecutionInfo().GetSearchAttributes().GetIndexedFields()["BuildIds"], &searchAttr)
+	s.NoError(err)
+	fmt.Printf("shahab> BuildIds=%v\n", searchAttr)
```
Comment (dnr) [inline tests/versioning.go:3751]: remove

File: service/history/workflow/mutable_state_impl.go [hunk 2]
```diff
@@ -2131,17 +2130,11 @@ func (ms *MutableStateImpl) UpdateBuildIdAssignment(buildId string) error {
 }
 
 func (ms *MutableStateImpl) updateBuildIdsSearchAttribute(version *commonpb.WorkerVersionStamp, maxSearchAttributeValueSize int) error {
```
Comment (dnr) [inline service/history/workflow/mutable_state_impl.go:2132]: rename `version` to `stamp` for consistency?

File: tests/versioning.go [hunk 3]
```diff
@@ -3736,15 +3741,24 @@ func (s *VersioningIntegSuite) validateWorkflowBuildId(
 	newVersioning bool,
 	expectingStamp bool,
 	expectedInheritedBuildId string,
+	extraSearchAttrBuildIds []string,
 ) {
 	dw, err := s.sdkClient.DescribeWorkflowExecution(ctx, wfId, runId)
 	s.NoError(err)
+	var searchAttr []string
+	err = converter.GetDefaultDataConverter().FromPayload(dw.GetWorkflowExecutionInfo().GetSearchAttributes().GetIndexedFields()["BuildIds"], &searchAttr)
```
Comment (dnr) [inline tests/versioning.go:3749]: use `searchattribute.DecodeValue(..., enumspb.INDEXED_VALUE_TYPE_KEYWORD_LIST, true)` like in MutableStateImpl.loadBuildIds?

File: service/history/workflow/mutable_state_impl.go [hunk 4]
```diff
@@ -2169,26 +2162,35 @@ func (ms *MutableStateImpl) loadBuildIds() ([]string, error) {
...
+// resulting SA list begins with either "unversioned" or "assigned:<bld>" based on workflow's Build ID assignment status.
+// Returns a potentially modified list.
+// [cleanup-old-wv] old versioning does not add "assigned:<bld>" value to the SA.
+func (ms *MutableStateImpl) addBuildIdToLoadedSearchAttribute(
+	existingValues []string,
+	stamp *commonpb.WorkerVersionStamp,
+) []string {
+	var newValues []string
+	if !stamp.GetUseVersioning() {
+		newValues = append(newValues, worker_versioning.UnversionedSearchAttribute)
```
[the remark this answers is not in the document]
Comment (ShahabT) [reply service/history/workflow/mutable_state_impl.go:2175]: it's used in VersionStampToBuildIdSearchAttribute, when stamp has build id but useVersioning is false.

File: service/history/workflow/mutable_state_impl.go [hunk 5]
```diff
@@ -2131,17 +2130,11 @@ func (ms *MutableStateImpl) UpdateBuildIdAssignment(buildId string) error {
...
 func (ms *MutableStateImpl) updateBuildIdsSearchAttribute(version *commonpb.WorkerVersionStamp, maxSearchAttributeValueSize int) error {
-	var toAdd []string
-	if !version.GetUseVersioning() {
-		toAdd = append(toAdd, worker_versioning.UnversionedSearchAttribute)
-	}
-	if version.GetBuildId() != "" {
-		toAdd = append(toAdd, worker_versioning.VersionStampToBuildIdSearchAttribute(version))
-	}
-	if len(toAdd) == 0 {
+	if version.GetBuildId() == "" {
```
[the remark this answers is not in the document]
Comment (ShahabT) [reply service/history/workflow/mutable_state_impl.go]: good catch. fixed

File: common/worker_versioning/worker_versioning.go [hunk 6]
```diff
@@ -41,19 +42,31 @@ import (
...
+		strings.HasPrefix(buildId, buildIdSearchAttributePrefixAssigned+BuildIdSearchAttributeDelimiter)
+}
+
 // VersionedBuildIdSearchAttribute returns the search attribute value for an unversioned build id
 func VersionedBuildIdSearchAttribute(buildId string) string {
 	return buildIdSearchAttributePrefixVersioned + BuildIdSearchAttributeDelimiter + buildId
 }
 
-// VersionedBuildIdSearchAttribute returns the search attribute value for an versioned build id
+// UnversionedBuildIdSearchAttribute returns the search attribute value for an versioned build id
```
Comment (carlydf) [inline common/worker_versioning/worker_versioning.go:69]:
```suggestion
// UnversionedBuildIdSearchAttribute returns the search attribute value for an unversioned build id
```

File: common/worker_versioning/worker_versioning.go [hunk 7]
```diff
@@ -41,19 +42,31 @@ import (
...
+	return buildIdSearchAttributePrefixAssigned + BuildIdSearchAttributeDelimiter + buildId
+}
+
+// IsUnversionedOrAssignedBuildIdSearchAttribute returns the value is "unversioned" or "assigned:<bld>"
+func IsUnversionedOrAssignedBuildIdSearchAttribute(buildId string) bool {
+	return buildId == UnversionedSearchAttribute ||
+		strings.HasPrefix(buildId, buildIdSearchAttributePrefixAssigned+BuildIdSearchAttributeDelimiter)
+}
+
 // VersionedBuildIdSearchAttribute returns the search attribute value for an unversioned build id
```
Comment (carlydf) [inline common/worker_versioning/worker_versioning.go:64]:
```suggestion
// VersionedBuildIdSearchAttribute returns the search attribute value for a versioned build id
```

**Review by carlydf — APPROVED** [review]
approved conditional on fixing the two comments that I proposed changes for

**Pull request merged**